### PR TITLE
[AIRFLOW-6885] Change delete-on-success to delete-on-failure

### PR DESCRIPTION
**Pull request by dimberman**
Jira: https://issues.apache.org/jira/browse/AIRFLOW-6885

It makes more sense to by-default not delete failed pods
Users should explicitly state they want these pods deleted.

This feature has not yet been released so ethis will not be a breaking
change

---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

File: airflow/config_templates/config.yml [hunk 1]
```diff
@@ -1752,9 +1752,9 @@
       type: string
       example: ~
       default: "True"
-    - name: delete_worker_pods_on_success
+    - name: delete_worker_pods_on_failure
       description: |
-        If True (default), worker pods will be deleted only on task success
+        If False (default), failed worker pods will not be deleted so users can investigate them.
```
Comment (ashb) [inline airflow/config_templates/config.yml]: You say False is the default, but True is :)

File: airflow/config_templates/default_airflow.cfg [hunk 2]
```diff
@@ -815,8 +815,8 @@ worker_container_image_pull_policy = IfNotPresent
 # If True, all worker pods will be deleted upon termination
 delete_worker_pods = True
 
-# If True (default), worker pods will be deleted only on task success
-delete_worker_pods_on_success = True
+# If False (default), failed worker pods will not be deleted so users can investigate them.
+delete_worker_pods_on_failure = True
```
Comment (ashb) [inline airflow/config_templates/default_airflow.cfg]: Just so I'm clear: if both of these delete_ options are set, then even failed pods will be deleted. Correct?

File: airflow/executors/kubernetes_executor.py [hunk 3]
```diff
@@ -895,11 +895,11 @@ def _change_state(self,
                       pod_id: str,
                       namespace: str) -> None:
         if state != State.RUNNING:
-            if self.kube_config.delete_worker_pods_on_success and state is State.SUCCESS:
+            if self.kube_config.delete_worker_pods and state is State.SUCCESS:
                 if not self.kube_scheduler:
                     raise AirflowException(NOT_STARTED_MESSAGE)
                 self.kube_scheduler.delete_pod(pod_id, namespace)
-            elif self.kube_config.delete_worker_pods:
+            elif self.kube_config.delete_worker_pods_on_failure:
```
Comment (ashb) [inline airflow/executors/kubernetes_executor.py]: How about changing the semantics a bit here: if we make `delete_worker_pods` be the "enable/disable deletion at all" and this new setting only controls what happens to failed pods:

```python
            if self.kube_config.delete_worker_pods:
               if not self.kube_scheduler:
                   raise AirflowException(NOT_STARTED_MESSAGE)
               if state is not State.FAILURE or self.delete_worker_pods_on_failure
                   self.kube_scheduler.delete_pod(pod_id, namespace)
```

And then the description of the config becomes:

```ini
# If True, worker pods will be deleted upon termination
delete_worker_pods = True

# If False (and delete_worker_pods is True), failed worker pods will not be deleted so users can investigate them.
delete_worker_pods_on_failure = True
```

**Review by ashb — COMMENTED** [review]
What do you think of this suggestion?

File: airflow/config_templates/default_airflow.cfg (hunk 2 above)
Comment (dimberman) [reply airflow/config_templates/default_airflow.cfg]: Yes. I'm making the DEFAULT to be "delete only succeeded pods."

File: airflow/executors/kubernetes_executor.py (hunk 3 above)
Comment (dimberman) [reply airflow/executors/kubernetes_executor.py]: Sure, LGTM

**Review by ashb — APPROVED** [review]
Was the original change against a Jira? We should probably use the same one in the subject of this one too.

**Review by kaxil — APPROVED** [review]

**Pull request merged**